xicam/SAXS/processing/correction.py

```diff
@@ -1,56 +1,118 @@
 
     def evaluate(self):
-        data = np.asarray(self.bitmasked_images.value.astype(np.uint16))
-        reduced_dark_image = self.dark_images.value
-        darks = None
-        if reduced_dark_image is not None:
-            reduced_dark_image = stackmean(np.asarray(self.dark_images.value, dtype=np.float32))
-            darks = np.array([reduced_dark_image,
-                              np.zeros(shape=reduced_dark_image.shape),
-                              np.zeros(shape=reduced_dark_image.shape)],
-                             dtype=np.float32)
-        self.corrected_images.value = correct_images(data,
-                                                     dark=darks)
-                                                     # flat=self.flat_field.value.astype(np.float32))
+        def correct(array, flats, bkg, gain_map=(1, 4, 8)):
+            # 16-bit unsigned
+            # image: bits 0 - 12
+            # bad:   bit 13
+            # gain:  bits 14-15
+            # gain1: 0b11 (3)
+            # gain2: 0b10 (2)
+            # gain8: 0b00 (0)
+            intensity = np.bitwise_and(0x1FFF, array)
+            bad_flag = np.bitwise_and(0x1, np.right_shift(array, 13))
+            gain = np.bitwise_and(0x3, np.right_shift(array, 14))
+            # map the gain bit values to the gain map indices to get the actual gain values
+            # e.g. 3 -> index 2; 2 -> index 1; 1 -> index 0
+            gain = np.vectorize(partial(lambda a, b: a[int((b + 1) / 2)], gain_map))(gain)
+            arr = flats * gain * intensity
+            arr = np.where(arr < bkg, bkg, arr)
+            return np.array((1 - bad_flag) * (arr - bkg), dtype=np.uint16)
+
+        if self.images.value.ndim != 3:
+            raise ValueError(f"\"images\" expects a 3-dimensional image array; shape = \"{self.images.value.shape}\"")
+
+        flats = self.flats.value
+        if flats is None:
+            flats = np.ones(self.images.value.shape[1:])
+        if flats.ndim != 2:
+            raise ValueError(f"\"flats\" should be 2-dimensional; shape = \"{self.flats.value.shape}\"")
+
+        darks = self.darks.value
+        if darks is None:
+            darks = np.zeros(self.images.value.shape)
+        if darks.ndim != 3:
+            raise ValueError(f"\"darks\" should be 3-dimensional; shape = \"{self.darks.value.shape}\"")
+
+        darks = np.sum(darks, axis=0) / darks.shape[0]
+        self.corrected_images.value = correct(np.asarray(self.images.value, dtype=np.uint16),
+                                              np.asarray(flats, dtype=np.float32),
+                                              np.asarray(darks, dtype=np.float32),
+                                              self.gains.value)
+
+
+try:
+    from csxtools.fastccd import correct_images
+    from csxtools.utils import stackmean
+    import numpy as np
+    from xicam.plugins.processingplugin import Input, Output, ProcessingPlugin
+
+
+    class CSXCorrectImage(ProcessingPlugin):
```

Comment:
If this class only exists for tests, can it be moved to the tests module?

Reply:
that makes sense, i went ahead and moved it there